### PR TITLE
Upgrade src/solver/hydro to C++

### LINKED ISSUE #1575: Remove malloc in src/solver/hydro/
Files are .c, upgrade this to .cpp and remove manual memory allocation

Code is duplicated between 
```
daily/h2o_j_resoudre_le_probleme_lineaire.cpp
daily2/h2o2_j_resoudre_le_probleme_lineaire.cpp

```

### PR DESCRIPTION
**Pull request by payetvin**
close #1575

### CODE REVIEW

**Review by a-zakir — APPROVED** [review]

**Comment by sylvlecl** [comment]
What about adding namespaces to those files while we are on the topic of migrating those to C++ ?

File: src/solver/hydro/daily/h2o_j_donnees_mensuelles.h [hunk 1]
```diff
@@ -34,39 +34,30 @@
...
 
-        /******************************************************************************************/
-        /* Problemes internes (utilise uniquement par l'optimisation) */
-        PROBLEME_HYDRAULIQUE* ProblemeHydraulique;
-    } DONNEES_MENSUELLES;
-
-#ifdef __cplusplus
-}
-#endif
+typedef struct
```
Comment (flomnes) [inline src/solver/hydro/daily/h2o_j_donnees_mensuelles.h]:
```suggestion
struct DONNEES_MENSUELLES
```

File: src/solver/hydro/daily/h2o_j_donnees_optimisation.h [hunk 2]
```diff
@@ -55,11 +58,11 @@ typedef struct
                                            */
     /* La matrice des contraintes */
     int NombreDeContraintes;
-    char* Sens;
-    int* IndicesDebutDeLigne;
-    int* NombreDeTermesDesLignes;
-    double* CoefficientsDeLaMatriceDesContraintes;
-    int* IndicesColonnes;
+    std::string Sens;
```
Comment (flomnes) [inline src/solver/hydro/daily/h2o_j_donnees_optimisation.h]: It's not really a string, more like `std::vector<char>`

File: src/solver/hydro/daily2/h2o2_j_donnees_mensuelles.h [hunk 3]
```diff
@@ -34,55 +34,46 @@
...
-        double* Turbine;       /* Resultat a recuperer par l'appelant */
-        double* niveauxFinJours;
-        double* overflows;
-        double* deviations;
-        double* violations;
-        double deviationMax;
-        double violationMax;
-        double waste;
+typedef struct
+{
```
Comment (flomnes) [inline src/solver/hydro/daily2/h2o2_j_donnees_mensuelles.h:42]: Don't use `typedef` here

**Review by flomnes — CHANGES_REQUESTED** [review]
There are still quite a few `free`, why ?

**Review by flomnes — APPROVED** [review]

**Pull request merged**